transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java

```diff
@@ -286,23 +329,21 @@ protected void run() {
                     default:
                         // fallthrough
                 }
-
+                lastBlockingAmount.lazySet(System.nanoTime() - runStartTime);
                 final int ioRatio = this.ioRatio;
+                final long ioStartTime = System.nanoTime();
+                lastProcessedChannels.lazySet(tryProcessReady(events, strategy));
+                final long ioEndTime = System.nanoTime();
+                final long ioTime = ioEndTime - ioStartTime;
                 if (ioRatio == 100) {
-                    if (strategy > 0) {
-                        processReady(events, strategy);
-                    }
                     runAllTasks();
                 } else {
-                    final long ioStartTime = System.nanoTime();
-
-                    if (strategy > 0) {
-                        processReady(events, strategy);
-                    }
-
-                    final long ioTime = System.nanoTime() - ioStartTime;
                     runAllTasks(ioTime * (100 - ioRatio) / ioRatio);
                 }
+                final long taskTime = System.nanoTime() - ioTime;
+                lastIoProcessingTime.lazySet(ioTime * 1000);
```

Comment:
I have implemented push based metrics/insights plugin for RxNetty and it does work well primarily because of it's flexibility to have different types of listeners, listening to the internal events. My motivation for providing a plugin mechanism was to not make decisions for the users as to how they want to store and publish this data (Much in line with what @Scottmitch is saying). In line with @ejona86 comment, it is a bit more complex due to API surface area and the awkwardness of co-relating between two different events. The awkwardness of co-relating events in RxNetty arise from the fact that the events are concurrent and span across multiple requests for example.

In the context of this change, I think the metric listener will be easier due to the following characteristics:

- It is specific to `EventLoop` so the API surface area would not be huge.
- The events will generally not be concurrent, specially for the ones where correlation is important (eg: time between `ioProcessingBegin()` and `ioProcessingEnd()`

So, I would 👍  the idea of a push based approach. My suggestion would be to not support multiple listeners (if that was a point of discussion).